### PR TITLE
fix(restart-pipeline) : CheckPrecondition doesn't evaluate expression correctly when upstream stages get restarted

### PR DESCRIPTION
**Pull request by rjalander**
This PR includes the changes to fix the [issue #6675](https://github.com/spinnaker/spinnaker/issues/6675) 

**Root cause:** 
Expression is not considered during restart of pipeline instead it picks the expression result from the previous run
**Fix:** 
Updating `StageExecution` for stage type `checkPreconditions` with expression from `restartDetails` Map
and store the updated `stage` with expression in the `ExecutionRepository`, from where the restart data gets picked.

This PR changes are dependent on https://github.com/spinnaker/gate/pull/1560

### CODE REVIEW

**Review by mattgogerly — CHANGES_REQUESTED** [review]

**Comment by mattgogerly** [comment]
Are there any tests you can add to validate that this fixes the problem you were seeing?

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/CompoundExecutionOperator.java [hunk 1]
```diff
@@ -119,6 +123,59 @@ public PipelineExecution restartStage(@Nonnull String executionId, @Nonnull Stri
...
+  }
+
+  private List<Map> getPreconditionsFromStage(Map restartDetails) {
+    List<Map> preconditionList = new ArrayList();
+    Map pipelineConfigMap = new HashMap(restartDetails);
+    List<String> keysToRetain = new ArrayList();
+    keysToRetain.add("stages");
+    pipelineConfigMap.keySet().retainAll(keysToRetain);
+    Map<String, List<Map>> pipelineStageMap = new HashMap(pipelineConfigMap);
+    if (!pipelineStageMap.isEmpty() && pipelineStageMap != null) {
```
[the remark this answers is not in the document]
Comment (rjalander) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/CompoundExecutionOperator.java]: Oh yeah, I missed it here.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/CompoundExecutionOperator.java [hunk 2]
```diff
@@ -119,6 +123,59 @@ public PipelineExecution restartStage(@Nonnull String executionId, @Nonnull Stri
...
+            }
+          }
+        }
+      }
+    }
+    return execution;
+  }
+
+  private List<Map> getPreconditionsFromStage(Map restartDetails) {
+    List<Map> preconditionList = new ArrayList();
```
[the remark this answers is not in the document]
Comment (rjalander) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/CompoundExecutionOperator.java:161]: Done

**Comment by rjalander** [comment]
@mattgogerly any idea how can we fix this check (Queue: Embarked in merge train) for merge.
I created the other PR (https://github.com/spinnaker/gate/pull/1560) with the same privileges and that PR passed this check (https://github.com/spinnaker/gate/runs/7301756832)

**Comment by andrew-est** [comment]
Adding comment so that I become a participant as I think that may let me have options as @Nordix user

**Comment by dbyron-sf** [comment]
That check is failing with this message:
image

**Comment by andrew-est** [comment]
The problem is we can't make the changes to follow the instructions and I think this is related to the issue raised in this discussion
https://github.com/github-community/community/discussions/5634

**Comment by mattgogerly** [comment]
If you're not able to make the changes to the config are you able to resubmit the PR from a different account?

**Pull request merged**